src/System.Net.Http/src/System/Net/Http/HPack/HPackDecoder.cs

```diff
@@ -0,0 +1,398 @@
+        //   0   1   2   3   4   5   6   7
+        // +---+---+---+---+---+---+---+---+
+        // | 0 | 0 | 1 |   Max size (5+)   |
+        // +---+---------------------------+
+        private const byte DynamicTableSizeUpdateMask = 0xe0;
+        private const byte DynamicTableSizeUpdateRepresentation = 0x20;
+
+        // http://httpwg.org/specs/rfc7541.html#rfc.section.5.2
+        //   0   1   2   3   4   5   6   7
+        // +---+---+---+---+---+---+---+---+
+        // | H |    String Length (7+)     |
+        // +---+---------------------------+
+        private const byte HuffmanMask = 0x80;
+
+        private const int IndexedHeaderFieldPrefix = 7;
+        private const int LiteralHeaderFieldWithIncrementalIndexingPrefix = 6;
+        private const int LiteralHeaderFieldWithoutIndexingPrefix = 4;
+        private const int LiteralHeaderFieldNeverIndexedPrefix = 4;
+        private const int DynamicTableSizeUpdatePrefix = 5;
+        private const int StringLengthPrefix = 7;
+
+        private readonly int _maxDynamicTableSize;
+        private readonly DynamicTable _dynamicTable;
+        private readonly IntegerDecoder _integerDecoder = new IntegerDecoder();
+        private readonly byte[] _stringOctets = new byte[MaxStringOctets];
+        private readonly byte[] _headerNameOctets = new byte[MaxStringOctets];
+        private readonly byte[] _headerValueOctets = new byte[MaxStringOctets];
+
+        private State _state = State.Ready;
+        private byte[] _headerName;
+        private int _stringIndex;
+        private int _stringLength;
+        private int _headerNameLength;
+        private int _headerValueLength;
+        private bool _index;
+        private bool _huffman;
+
+        public HPackDecoder(int maxDynamicTableSize = DefaultHeaderTableSize)
+            : this(maxDynamicTableSize, new DynamicTable(maxDynamicTableSize))
+        {
+            _maxDynamicTableSize = maxDynamicTableSize;
+        }
+
+        // For testing.
+        internal HPackDecoder(int maxDynamicTableSize, DynamicTable dynamicTable)
+        {
+            _maxDynamicTableSize = maxDynamicTableSize;
+            _dynamicTable = dynamicTable;
+        }
+
+        public void Decode(ReadOnlySpan<byte> data, HeaderCallback onHeader)
+        {
+            for (var i = 0; i < data.Length; i++)
+            {
+                OnByte(data[i], onHeader);
+            }
+        }
+
+        public void CompleteDecode()
+        {
+            if (_state != State.Ready)
+            {
+                throw new HPackDecodingException("CoreStrings.HPackErrorIncompleteHeaderBlock");
+            }
+        }
+
+        private void OnByte(byte b, HeaderCallback onHeader)
+        {
+            switch (_state)
+            {
+                case State.Ready:
+                    if ((b & IndexedHeaderFieldMask) == IndexedHeaderFieldRepresentation)
+                    {
+                        var val = b & ~IndexedHeaderFieldMask;
```

Comment:
Nit: please revise any such `var`s... I'll avoid commenting on all of them :)

Reply:
fixing